File: backend/app/schemas/direccion.py
"""Schemas del área de dirección."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class GrupoCreate(BaseModel):
    tipo: str = Field(min_length=1, max_length=24)
    leader: str | None = Field(default=None, max_length=120)
    channel: str | None = Field(default=None, max_length=60)
    members_total: int = 0
    members_active: int = 0


class GrupoRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    tipo: str
    estado: str
    members_total: int
    members_active: int
    leader: str | None
    channel: str | None


class SolicitudMediosCreate(BaseModel):
    requested_by: str = Field(min_length=1, max_length=120)
    resource_type: str = Field(min_length=1, max_length=40)
    quantity: int = Field(default=1, ge=1)
    reason: str | None = Field(default=None, max_length=500)
    priority: str = Field(default="medium", pattern="^(critical|high|medium|low)$")


class SolicitudMediosRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    requested_by: str
    resource_type: str
    quantity: int
    reason: str | None
    priority: str
    estado: str
    requested_at: datetime
    decided_at: datetime | None


class ComunicadoCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    body: str = Field(min_length=1, max_length=4000)
    audience: str = Field(default="internal", pattern="^(internal|press|population|authorities)$")


class ComunicadoRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    title: str
    body: str
    audience: str
    estado: str
    created_by: str | None
    created_at: datetime


class AlbergueCreate(BaseModel):
    name: str = Field(min_length=1, max_length=160)
    lat: float | None = None
    lng: float | None = None
    capacity: int = Field(default=0, ge=0)


class AlbergueRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: int
    emergencia_id: int
    name: str
    lat: float | None
    lng: float | None
    capacity: int
    occupancy: int
    estado: str


class EstadoUpdate(BaseModel):
    estado: str = Field(min_length=1, max_length=20)


class OcupacionUpdate(BaseModel):
    occupancy: int = Field(ge=0)

File: frontend/src/types/direccion.ts
import type { GeoPoint, ID, ISODateString, Severity } from "./common";
import type { ResourceKind } from "./emergency";
import type { OperationalState } from "./operational";

// Operational level of the activated plan (Spanish civil protection convention).
export type OperationalLevel = 0 | 1 | 2 | 3;

export const OPERATIONAL_LEVEL_LABEL: Record<OperationalLevel, string> = {
  0: "PRE-EMERGENCIA",
  1: "NIVEL 1 — LOCAL",
  2: "NIVEL 2 — PROVINCIAL",
  3: "NIVEL 3 — ESTATAL",
};

// Functional groups under the Director del Plan.
export type OperationalGroup =
  | "command"        // CECOPAL / Mando
  | "intervention"   // Intervención (Bomberos)
  | "medical"        // Sanitario
  | "security"       // Seguridad / FCSE
  | "logistics"      // Logístico
  | "technical"      // Apoyo Técnico
  | "psychosocial";  // Psicosocial

export const GROUP_LABEL: Record<OperationalGroup, string> = {
  command: "MANDO · CECOPAL",
  intervention: "INTERVENCIÓN",
  medical: "SANITARIO",
  security: "SEGURIDAD",
  logistics: "LOGÍSTICO",
  technical: "APOYO TÉCNICO",
  psychosocial: "PSICOSOCIAL",
};

export const GROUP_SHORT: Record<OperationalGroup, string> = {
  command: "MND",
  intervention: "INT",
  medical: "SAN",
  security: "SEG",
  logistics: "LOG",
  technical: "TEC",
  psychosocial: "PSI",
};

export interface GroupStatus {
  id: ID;
  type: OperationalGroup;
  state: OperationalState;
  membersTotal: number;
  membersActive: number;
  leader: string;
  channel?: string;          // e.g. "TETRA-104"
  lastReport: ISODateString;
}

// Solicitud de medios (PMA → Director / CECOPAL)
export type MediaRequestStatus = "pending" | "approved" | "denied" | "delivered";

export interface MediaRequest {
  id: ID;
  emergencyId: ID;
  requestedBy: string;       // PMA-Norte-01
  resourceType: ResourceKind | "personnel" | "supplies" | "transport";
  quantity: number;
  reason: string;
  priority: Severity;
  status: MediaRequestStatus;
  requestedAt: ISODateString;
  decidedAt?: ISODateString;
  eta?: ISODateString;
}

// Comunicado público o interno
export type CommuniqueAudience = "internal" | "press" | "population" | "authorities";
export type CommuniqueStatus = "draft" | "pending-approval" | "approved" | "sent";

export interface Communique {
  id: ID;
  emergencyId: ID;
  title: string;
  body: string;
  audience: CommuniqueAudience;
  status: CommuniqueStatus;
  createdBy: string;
  createdAt: ISODateString;
  approvedAt?: ISODateString;
  sentAt?: ISODateString;
}

// Evacuación (zona origen → albergue)
export type EvacuationStatus = "planned" | "in-progress" | "completed";

export interface Evacuation {
  id: ID;
  emergencyId: ID;
  name: string;
  status: EvacuationStatus;
  toShelterId: ID;
  estimatedPeople: number;
  evacuatedPeople: number;
  routePoints: GeoPoint[];   // polyline hacia albergue
  startedAt?: ISODateString;
}

// Albergue / centro de acogida
export interface Shelter {
  id: ID;
  name: string;
  location: GeoPoint;
  capacity: number;
  occupancy: number;
  state: OperationalState;
  contact?: string;
  facilities?: string[];
}

// Corte de vía / acceso
export type RoadBlockStatus = "active" | "intermittent" | "lifted";

export interface RoadBlock {
  id: ID;
  emergencyId: ID;
  road: string;
  km?: string;
  location: GeoPoint;          // midpoint (compatibility)
  segment?: { from: GeoPoint; to: GeoPoint };  // calle cortada de A a B
  reason: string;
  status: RoadBlockStatus;
  createdAt: ISODateString;
}

// Puesto de Mando Avanzado / CECOPAL en mapa
export type CommandPostType = "PMA" | "CECOPAL" | "PCA";

export interface CommandPost {
  id: ID;
  emergencyId: ID;
  code: string;
  type: CommandPostType;
  location: GeoPoint;
  commanderName: string;
  state: OperationalState;
  groups: OperationalGroup[];
  installedAt: ISODateString;
}

export interface DirectorAction {
  id: ID;
  emergencyId: ID;
  type:
    | "activated"
    | "level-escalated"
    | "level-deescalated"
    | "communique-approved"
    | "support-requested"
    | "closed";
  performedBy: string;
  timestamp: ISODateString;
  payload?: Record<string, unknown>;
  notes?: string;
}
